### ISSUE TITLE
Error: Unable to open the Yubikey card: connecting to smart card: the smart card cannot be accessed because of other connections outstanding

### ISSUE THREAD

**Issue by dud225** [go-piv/piv-go#116]
Hello

I'm able to build a program with this library that allows me to successfully interact with my YubiKey. However under some conditions, e.g. after some idle period, I'm getting the aforementioned erorr:
```
Error: Unable to open the Yubikey card: connecting to smart card: the smart card cannot be accessed because of other connections outstanding
```

I'm not sure if there is anything wrong with piv-go. But I'm quite puzzled since I make thorough use of the `Close()` method and additionally I don't get this error with other tools e.g. libykcs11.so.

**Comment by udf2457** [comment]
I stumbled accross this one today too.

I am using `Close()` but this does not appear to be honoured.

Even if I add `time.Sleep(20 * time.Second)` after `Close()` , `piv-go` still seems to be holding on to that connection.

**Comment by dstrbad** [comment]
Any update on this one maybe? We stumbled across this one as well.

**Comment by andsens** [comment]
From https://github.com/FiloSottile/yubikey-agent/issues/98#issuecomment-931785647  
Use `sudo lsof +E /run/pcscd/pcscd.comm` to see which processes may be accessing the yubikey. In my case it was gnomes `gsd-smartcard` daemon.

**Comment by quite** [comment]
I find the following a little odd. When the setup is in such a state that a program that uses go-piv fails with the above message, it seems the program `yubico-piv-tool` can still be run to do for example `-a status`, and `-s 82 -a read-certificate`. This program also links with `libpcsclite.so.1` (like a Go program using go-piv), so I assume it is using the same methods? But perhaps not the exact same calls, or so? Hm

**Comment by DavidBuchanan314** [comment]
Yes, definitely something weird going on. I do see `gsd-smartcard` in lsof, but even if I kill the process I still get the `other connections outstanding` error. (edit from the future: there are still other processes using it, including my browser and gnome itself. I can't reasonably kill them all)

All the while, `yubico-piv-tool` works just fine.

Some ltrace logs:

```
SCardEstablishContext(2, 0, 0, 0x4000192380)                            = 0
SCardListReaders(0x1c127ad5, 0, 0, 0x4000192390)                        = 0
SCardListReaders(0x1c127ad5, 0, 0x40001b6100, 0x4000192390)             = 0
SCardReleaseContext(0x1c127ad5, 0x5d1ccc, 0x40001f9550, 0x859160)       = 0
SCardEstablishContext(2, 0, 0, 0x4000192398)                            = 0
malloc(31)                                                              = 0x19b3a1f0
SCardConnect(0x73ef254e, 0x19b3a1f0, 1, 2)                              = 0x8010000b
SCardReleaseContext(0x73ef254e, 0x5d1ccc, 0x40001f9570, 0x859160)       = 0
```
Where `0x8010000b` is the error code for the outstanding connections error.


(can't seem to ltrace yubico-piv-tool properly for some reason)

Ah, just needed a `-x "@libpcsclite.so.1"`. Here's corresponding ltrace from `yubico-piv-tool`

```
SCardIsValidContext@libpcsclite.so.1(-1, 0, 0xffffded82cf0, 0xffffded83480) = 0x80100001
SCardEstablishContext@libpcsclite.so.1(2, 0, 0, 0xaaab4611dec0)         = 0
SCardListReaders@libpcsclite.so.1(0x68a218b2, 0, 0xffffded82cf8, 0xffffded82c90) = 0
SCardConnect@libpcsclite.so.1(0x68a218b2, 0xffffded82cf8, 2, 3)         = 0
SCardStatus@libpcsclite.so.1(0x4a7da556, 0xffffded82070, 0xffffded82068, 0) = 0
SCardBeginTransaction@libpcsclite.so.1(0x4a7da556, 0, 0, 0)             = 0
SCardTransmit@libpcsclite.so.1(0x4a7da556, 0xfffed2e01e58, 0xffffded7fc08, 11) = 0
SCardTransmit@libpcsclite.so.1(0x4a7da556, 0xfffed2e01e58, 0xffffded7ee88, 5) = 0
SCardTransmit@libpcsclite.so.1(0x4a7da556, 0xfffed2e01e58, 0xffffded7fac8, 5) = 0
SCardTransmit@libpcsclite.so.1(0x4a7da556, 0xfffed2e01e58, 0xffffded7f988, 5) = 0
...
```

Some observations on differences:

- piv-go appears to establish and release a context just for listing readers, and then establishes a new one for the connection. Probably not relevant.
- third and fourth params for SCardConnect differ. Looking at <https://pcsclite.apdu.fr/api/group__API.html#ga4e515829752e0a8dbc4d630696a8d6a5> I'd guess they're the difference between `SCARD_SHARE_SHARED` and `SCARD_SHARE_EXCLUSIVE`

Will dig in on that second point.

Yes, that's exactly it https://github.com/go-piv/piv-go/blob/102a86cf189e62ae83632502af65b8c0bf0abb1a/v2/piv/pcsc_unix.go#L101

https://github.com/Yubico/yubico-piv-tool/blob/73db815e8927028c51ba71771a6737efaa238a62/lib/ykpiv.c#L955

So the question is, do we really need `SCARD_SHARE_EXCLUSIVE`, given that yubico-piv-tool appears to use `SCARD_SHARE_SHARED` just fine? Could we add an option for it?

**Comment by DavidBuchanan314** [comment]
Can confirm that `SCARD_SHARE_SHARED` seems to be working fine for my use-case
